src/jit/optimizer.cpp

```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
                 }
 #endif // FEATURE_EH_FUNCLETS && defined(_TARGET_ARM_)
 
-                /* At this point we have a loop - record it in the loop table
+                // Compact the loop (sweep through it and move out any blocks that aren't part of the
+                // flow cycle), and count the exits.
+                BasicBlock* moveAfter = nullptr;
+                for (BasicBlock* previous = top->bbPrev; previous != bottom;)
+                {
+                    BasicBlock* block = previous->bbNext;
```

Comment:
I agree the function is more monolithic than ideal, so I'll see if I can find an appropriate way to refactor, but at first blush I don't think it would end up reading better, given the amount of state that would need to get passed in and particularly that would need to get passed out, not to mention I'd need to pull the `LoopBlockSet` definition out of line.